answerking_app/utils/mixins/RetireMixin.py

```diff
@@ -1,9 +1,61 @@
-from rest_framework.mixins import UpdateModelMixin
+from django.db.models import QuerySet
+from rest_framework import status
+from rest_framework.generics import GenericAPIView
 from rest_framework.request import Request
 from rest_framework.response import Response
+from rest_framework.utils.serializer_helpers import ReturnDict
 
+from answerking_app.models.models import Category, Product, Order, OrderLine
+from answerking_app.models.serializers import (
+    CategorySerializer,
+    ProductSerializer,
+    OrderSerializer,
+)
+from answerking_app.utils.mixins.ApiExceptions import HttpErrorResponse
+from answerking_app.utils.model_types import OrderProductType
 
-class RetireMixin(UpdateModelMixin):
+
+class RetireMixin(GenericAPIView):
     def retire(self, request: Request, *args, **kwargs) -> Response:
-        request.data["retired"] = True
-        return super().partial_update(request, *args, **kwargs)
+        instance: Category | Product = self.get_object()
+        if instance.retired:
+            raise HttpErrorResponse(
+                status=status.HTTP_410_GONE,
+                detail="This object has already been retired",
+            )
+        if isinstance(instance, Category):
+            instance.retired = True
+            instance.save()
+            response: ReturnDict = CategorySerializer(instance).data
+        else:
```

Comment:
what if instance is neither a Category or Product? think you should have an elif for product and an else which raises an error.

Reply:
yes currently, but this could change down the line. And while we only use it in the intended way, if someone comes back in a years time and missuses it it will not error it will alter the code in an undesired way and do it silently